add downward api

DennisDenuto opened the pull request:
#### What this PR does / why we need it:
Draft PR WIP. Still need to:

- [x] clean up commits
- [x]  fix pb gen code
- [x]  add tests around annotations / labels
- [x]  add test coverage for other non-ytt templating
- [x]  add comments
- [x]  do the TODO's
- [x]  think about allowing configuring a path as a valid option
- [x]  fix up commits / messages
- [ ]  split out doc change (dev.md) into a separate PR
- [x]  Add documentation to carvel.dev as a separate PR

#### Which issue(s) this PR fixes:

Fixes #659 

#### Does this PR introduce a user-facing change?
```release-note
Allow specifying the App name, namespace, uid, annotations and labels as values during templating
```

#### Additional Notes for your reviewer:

##### Review Checklist:


- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

DennisDenuto committed:
feat: add downward api

- name, namespace, annotations, labels and uid are supported fields
- add generated pb

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
remove unused code

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/template/values.go:
```diff
@@ -102,6 +110,105 @@ func (t Values) writeFromSecret(dstPath string,
+	return result, nil
+}
+
+func (t Values) extractFieldPathAsKeyValue(name string, fieldPath string) ([]byte, error) {
```
wdyt about using json path library that kubectl uses for extracting values?

DennisDenuto replied:
yeah thx for the pointer! quick search shows up https://github.com/kubernetes/kubectl/blob/ac26f503e81287d9903761a1a8ded25fdebec6a7/pkg/cmd/get/customcolumn.go#L47

I'll def have a closer look as I can see this being a more generalized solution.

Should I just go ahead and load up the entire metadata and allow folk to inject any jsonpath within '.metadata'? or do we still want to be restrictive to only allow name, namespace, labels, annotations, uid as valid json paths?

cppforlife replied:
looks like it's using https://github.com/kubernetes/client-go/tree/master/util/jsonpath.

i think it's still probably a good idea to limit, but instead of limiting the json path, we can limit input structure. that way you dont have make assertions against a query.

DennisDenuto committed:
refactor: rename GenericOpts to AppContext

- Hope is for the new name to better capture the purpose and usage of
  the data struct

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
test refactor: template_test

- adds e2e happy cases for downward api
- add template test for helm
- re-order the tests, grouping the "ValuesFrom" tests for ytt, cue and
  helm together

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
feat: allow nested keys in downward api

- validate keys provided by operator
- allows keys to contain '.' which create a nested yaml map
-- keys with a '.' can still escape the '.' with '\.'

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/apis/kappctrl/v1alpha1/types_template.go:
```diff
@@ -57,16 +57,29 @@ type AppTemplateHelmTemplate struct {
+type DownwardAPIAppTemplateValues struct {
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	// Required: Selects a field of the app: only annotations, labels, uid, name and namespace are supported.
+	FieldPath string `json:"fieldPath" protobuf:"bytes,2,opt,name=fieldPath"`
```
probably should add omitempty here as well?

DennisDenuto replied:
done

cppforlife commented on pkg/template/helm_template.go:
```diff
@@ -109,8 +109,3 @@ func (t *HelmTemplate) template(dirPath string, input io.Reader) exec.CmdRunResu
 	return result
 }
-
-// auxiliary struct used for Chart.yaml unmarshalling
```
nice!

cppforlife commented on pkg/template/interfaces.go:
```diff
@@ -23,7 +25,26 @@ type Template interface {
+type AppContext struct {
 	Name      string
 	Namespace string
+	Metadata  *PartialObjectMetadata
```
do we need this to be a pointer?

DennisDenuto replied:
don't think so. Removed pointer

cppforlife commented on pkg/template/values.go:
```diff
@@ -102,10 +112,104 @@ func (t Values) writeFromSecret(dstPath string,
 	return result, nil
 }
 
+func (t Values) writeFromDownwardAPI(dstPath string, downwardAPIRef v1alpha1.AppTemplateValuesDownwardAPI) ([]string, error) {
```
feels like we should extract this method and associated methods below into ValuesDownwardAPI class. seems like Values is growing a bit too much. (unit tests can stay the same though)

DennisDenuto replied:
Yes! +1 

I went back and forth on adding a new `ValuesDownwardAPI` to encapsulate this logic. I landed on "let's see if someone else thinks this during the review"

:-)

cppforlife commented on pkg/template/values.go:
```diff
@@ -102,10 +112,104 @@ func (t Values) writeFromSecret(dstPath string,
+
+func (Values) validateName(name string) error {
+	if strings.HasSuffix(name, ".") || strings.HasPrefix(name, ".") || strings.Contains(name, "..") {
+		return errors.New("Invalid name was provided (hint: separate paths should only use a single '.' character)")
```
probably best to include name in the error.

DennisDenuto replied:
ok done

cppforlife commented on pkg/template/values.go:
```diff
@@ -102,10 +112,104 @@ func (t Values) writeFromSecret(dstPath string,
+	}
+
+	if len(results) != 1 || len(results[0]) != 1 {
+		return nil, errors.New("invalid field spec provided to DownwardAPI. Only single supported fields are allowed")
```
nit: capital `I`

DennisDenuto replied:
attention 👏 to 👏 detail 👏

cppforlife commented on pkg/template/values.go:
```diff
@@ -102,10 +112,104 @@ func (t Values) writeFromSecret(dstPath string,
+func (Values) nestedKey(name string) string {
+	var nested string
+	for idx, n := range splitWithEscaping(name, '.', '\\') {
+		nested += fmt.Sprintf("%s%s:\n", strings.Repeat("  ", idx), n)
```
instead of building up yaml content via strings, lets use maps and yaml.Marshal the result.

DennisDenuto replied:
ok done

cppforlife commented on pkg/template/values.go:
```diff
@@ -102,10 +112,104 @@ func (t Values) writeFromSecret(dstPath string,
+// operator may wish to assign a downward API value into a nested key structure to use within their template
+func (Values) nestedKey(name string) string {
+	var nested string
+	for idx, n := range splitWithEscaping(name, '.', '\\') {
```
im trying to think, we have any prior art in escaping value keys... in carvel/kapp-controller cc @pivotaljohn

cppforlife replied:
related conversations in ytt: https://github.com/vmware-tanzu/carvel-ytt/issues/671

pivotaljohn replied:
LGTM... IMO, this is actually the ideal UX for ytt users.  So, when we implement 671, we'll also use the slash for escaping dots in paths. 👍🏻

DennisDenuto committed:
refactor:

- encapsualte DownwardAPI values logic into DownwardAPIValues
- instead of building up yaml content via strings, use maps and yaml.Marshal the result.

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/apis/kappctrl/v1alpha1/types_template.go:
```diff
@@ -57,16 +57,29 @@ type AppTemplateHelmTemplate struct {
 }
 
+// +k8s:openapi-gen=true
+type DownwardAPIAppTemplateValues struct {
```
nit: i would recommend calling this `AppTemplateValuesDownwardAPIItem` so that it matches parent `AppTemplateValuesDownwardAPI`

DennisDenuto replied:
ok done

cppforlife commented on pkg/template/metadata_values_extractor.go:
```diff
@@ -0,0 +1,176 @@
+		return a.keyValues(name, result.Interface().(map[string]string))
+	}
+
+	yamlContent, _, err := a.nestedKeyValue(name, fmt.Sprintf("%v", result.Interface()))
```
to get rid of all string building use `a.nestedKeyValue(name, result.Interface())`.

also no need to check on `if kind == reflect.Map {` above.

DennisDenuto replied:
if map is provided to `a.nestedKeyValue(name, map)` then the yaml produced will not be ordered / consistent across multiple templating.

Is that ok?

DennisDenuto replied:
Actually, yaml.Marshall seems to also provide stable out. so I just went ahead and did that! :-)

cppforlife commented on pkg/template/values.go:
```diff
@@ -102,27 +109,28 @@ func (t Values) writeFromSecret(dstPath string,
 
-	configMap, err := t.coreClient.CoreV1().ConfigMaps(t.genericOpts.Namespace).Get(
-		context.Background(), configMapRef.Name, metav1.GetOptions{})
+func (t Values) writeFromDownwardAPI(dstPath string, valuesExtractor DownwardAPIValuesExtractor) ([]string, error) {
```
can just take a concrete type especially since method is called writeFromDownwardAPI.

DennisDenuto replied:
ok done

cppforlife commented on pkg/template/metadata_values_extractor.go:
```diff
@@ -0,0 +1,176 @@
+
+// ObjectMetadataValuesExtractor produces multiple key-values based on the DownwardAPI config
+// queried against the object metadata
+type ObjectMetadataValuesExtractor struct {
```
wdyt about just calling this: DownwardAPIValues

DennisDenuto replied:
ok changed

cppforlife commented on pkg/template/metadata_values_extractor.go:
```diff
@@ -0,0 +1,176 @@
+}
+
+// Values returns many key-values queried (using jsonpath) against an object metadata provided.
+func (a ObjectMetadataValuesExtractor) Values() ([][]byte, error) {
```
i would recommend calling this `AsYAMLs()`

DennisDenuto replied:
ok changed

cppforlife commented on test/e2e/kappcontroller/template_test.go:
```diff
@@ -113,6 +117,18 @@ spec:
+              fieldPath: metadata.name
+            - name: uid
+              fieldPath: metadata.uid
+            - name: annotation
+              fieldPath: metadata.annotations['expectedAnnotation']
```
can we also throw item that returns a map

```
- name: annotations
  fieldPath: metadata.annotations
```

DennisDenuto replied:
ok will do.
